server/src/main/java/org/jboss/as/server/deployment/module/descriptor/DeploymentStructureDescriptorParser.java

```diff
@@ -238,9 +239,20 @@ public void deploy(final DeploymentPhaseContext phaseContext) throws DeploymentU
         }
     }
 
-    private void handleDeployment(final DeploymentPhaseContext phaseContext, final DeploymentUnit deploymentUnit, final ModuleSpecification moduleSpec, final ModuleStructureSpec rootDeploymentSpecification, Set<ModuleIdentifier> additionalModules) throws DeploymentUnitProcessingException {
+    private void handleDeployment(final DeploymentPhaseContext phaseContext, final DeploymentUnit deploymentUnit, final ModuleSpecification moduleSpec, final ModuleStructureSpec rootDeploymentSpecification, final Set<AdditionalModuleSpecification> additionalModules) throws DeploymentUnitProcessingException {
         final Map<VirtualFile, ResourceRoot> resourceRoots = resourceRoots(deploymentUnit);
-        moduleSpec.addUserDependencies(rootDeploymentSpecification.getModuleDependencies());
+        // handle unmodifiable module dependencies with alias
+        List<ModuleDependency> moduleDependencies = rootDeploymentSpecification.getModuleDependencies();
+        List<ModuleDependency> aliasDependencies = new ArrayList<>();
+        for (ModuleDependency dependency : moduleDependencies) {
+            ModuleIdentifier identifier = dependency.getIdentifier();
+            for (AdditionalModuleSpecification module : additionalModules) {
+                if (module.getAliases().contains(identifier)) {
```

Comment:
Is module.getAliases() as Set as well?  If not, this contains operation could get somewhat expensive in some cases.

Reply:
A Map<ModuleIdentifier, AdditionalModuleSpecification> doesn't help to eliminate it. It doesn't compare with map key, it needs to compare module dependency's identifier with each map value additional module's aliases something like:
```
List<ModuleDependency> moduleDependencies = rootDeploymentSpecification.getModuleDependencies();
        List<ModuleDependency> aliasDependencies = new ArrayList<>();
        for (ModuleDependency dependency : moduleDependencies) {
            ModuleIdentifier identifier = dependency.getIdentifier();
            for (AdditionalModuleSpecification module : additionalModules.values()) {
                if (module.getAliases().contains(identifier)) {
                    aliasDependencies.add(new ModuleDependency(dependency.getModuleLoader(), module.getModuleIdentifier(), dependency.isOptional(), dependency.isExport(), dependency.isImportServices(), dependency.isUserSpecified()));
                }
            }
        }
```

Reply:
Hi @dmlloyd IIUC, we need to compare module dependency identifier with additional module aliases. A map with ModuleIdentifier as key does not eliminate the linear searches.

Reply:
You're iterating all of `additionalModules.values()` for every single value in `moduleDependencies` looking for the given `dependency`, and you're iterating every alias looking for `identifier`.  To reduce a triply-nested loop to a doubly-nested loop, you could write it like this:

```java
        List<ModuleDependency> moduleDependencies = rootDeploymentSpecification.getModuleDependencies();
        List<ModuleDependency> aliasDependencies = new ArrayList<>();
        // Pre-index additionalmodules
        Map<ModuleIdentifier, ModuleDependency> index = new HashMap<>();
        // Only iterate additionalModules once, instead of once per dependency
        for (AdditionalModuleSpecification module : additionalModules.values()) {
            for (ModuleIdentifier alias : module.getAliases()) {
                index.put(alias, module);
            }
        }
        // No more nested loop
        for (ModuleDependency dependency : moduleDependencies) {
            ModuleIdentifier identifier = dependency.getIdentifier();
            if (index.containsKey(identifier)) {
                aliasDependencies.add(new ModuleDependency(dependency.getModuleLoader(), module.getModuleIdentifier(), dependency.isOptional(), dependency.isExport(), dependency.isImportServices(), dependency.isUserSpecified()));
            }
        }
```

The `index` could probably even be a simple `Set` because it doesn't look as if the corresponding value is even used by the existing code (you might want to double-check that for correctness because it looks a little wrong to me).

Reply:
Thanks for the review and kind explanation
The corresponding values in map are needed in order to obtain real ModuleIdentifier so we can add it into aliasDependencies like:

```
            if (index.containsKey(identifier)) {
                aliasDependencies.add(new ModuleDependency(dependency.getModuleLoader(), index.get(identifier).getModuleIdentifier(), dependency.isOptional(), dependency.isExport(), dependency.isImportServices(), dependency.isUserSpecified()));
            }
```

I pushed the change again, please review at your convenience. Once this is approved,  I will squash commits together. Thanks.

Reply:
Looks OK to me.